docs/tutorials/bayc/step-two-deriving-owners-and-tokens.md

```diff
@@ -0,0 +1,291 @@
+     blockNumber: Int! @index
+     blockTimestamp: DateTime! @index
+     transactionHash: String! @index
+-    from: String! @index
+-    to: String! @index
+-    tokenId: BigInt! @index
++    from: Owner!
++    to: Owner!
++    token: Token!
+}
+```
+
+Finally, add the virtual [reverse lookup fields](/basics/schema-file/entity-relations/):
+```diff
+ type Owner @entity {
+     id: ID! # owner address
++    ownedTokens: [Token!]! @derivedFrom(field: "owner")
+ }
+
+ type Token @entity {
+     id: ID! # string form of tokenId
+     tokenId: Int!
++    transfers: [Transfer!]! @derivedFrom(field: "token")
+ }
+```
+This adds no database columns, but makes `ownedTokens` and `transfers` fields available via GraphQL and Typeorm.
+
+The final version of `schema.graphql` is available [here](https://github.com/abernatskiy/tmp-bayc-squid/blob/c4bc82a9625d0c82b1bb5166e76add46099ac9e8/schema.graphql). Regenerate the Typeorm entity code once you're done:
+```bash
+sqd codegen
+```
+
+## An architecture for entity generation
+
+Note how the entities we define form an acyclic dependency graph:
+- `Owner` entity instances are generated straight from the raw `Transfer` events data;
+- `Token`s require the raw data plus the `Owner`s;
+- `Transfer` entities require all of the above.
+
+As a consequence, entity generation must proceed in a [particular order](https://en.wikipedia.org/wiki/Topological_sorting). Squids usually use small graphs like this one, and in these the order can be easily found manually (e.g. `Owner`s then `Token`s then `Transfer`s in this case). Hence, we will not automate the search for the order, but will assume that it can be hardcoded by the programmer.
+
+Some of the steps of entity generation may involve database lookups, IPFS retrieval or some other operations that take less time when done in batches. Hence, we will need some place to hold the raw data collected from the [whole batch of blocks](/basics/batch-processing/) and any previously generated entities that the new ones may depend on. For convenience, let's make an object that does that and also holds the functions that perform each entity generation step, as well as the correct order for calling them.
+
+Let us repurpose the existing `EntityBuffer` singleton class to do that. Begin by renaming it:
+```bash
+sed -i -e 's/EntityBuffer/EntityGenerator/g' ./src/entityBuffer.ts ./src/mapping/contract.ts ./src/processor.ts
+mv src/entityBuffer.ts src/entityGenerator.ts
+sed -i -e 's/entityBuffer/entityGenerator/g' ./src/mapping/contract.ts ./src/processor.ts
+```
+The new class should hold
+- raw blockchain data;
+- entities;
+- callback functions that create new entities given the raw data and the previously generated entities;
+- order of entity generation.
+
+Adding this and a few convenience methods, we arrive at
+```typescript
+import assert from 'assert'
+
+interface Entity {
+    id: string
+    constructor: {
+        name: string
+    }
+}
+
+export class EntityGenerator {
```

Comment:
So why exactly do we need EntityGenerator?

Reply:
Looks too complex for such simple project